extensions/sql-migration/src/service/serviceClient.ts

```diff
@@ -0,0 +1,255 @@
+	}
+
+	private createClientOptions(): ClientOptions {
+		return {
+			providerId: constants.providerId,
+			errorHandler: new LanguageClientErrorHandler(),
+			synchronize: {
+				configurationSection: [constants.extensionConfigSectionName, constants.sqlConfigSectionName]
+			},
+			features: [
+				// we only want to add new features
+				SqlMigrationService,
+				TdeMigrationService
+			],
+			outputChannel: new CustomOutputChannel()
+		};
+	}
+
+	private generateServerOptions(executablePath: string, context: vscode.ExtensionContext): ServerOptions {
+		let launchArgs = [];
+		launchArgs.push(`--locale`, vscode.env.language);
+		launchArgs.push('--log-file', path.join(context.logUri.fsPath));
+		launchArgs.push('--tracing-level', this.getConfigTracingLevel());
+		launchArgs.push('--autoflush-log');
+		return { command: executablePath, args: launchArgs, transport: TransportKind.stdio };
+	}
+
+	private getConfigTracingLevel(): TracingLevel {
+		let config = vscode.workspace.getConfiguration('mssql');
+		if (config) {
+			return config['tracingLevel'];
+		} else {
+			return TracingLevel.Critical;
+		}
+	}
+
+	private generateHandleServerProviderEvent(): EventAndListener {
+		let dots = 0;
+		return (e: string | string[], ...args: any[]) => {
+			switch (e) {
+				case Events.INSTALL_START:
+					this.outputChannel.show(true);
+					this.statusView.show();
+					this.outputChannel.appendLine(localize('installingServiceDetailed', "Installing {0} to {1}", constants.serviceName, args[0]));
+					this.statusView.text = localize('installingService', "Installing {0} Service", constants.serviceName);
+					break;
+				case Events.INSTALL_END:
+					this.outputChannel.appendLine(localize('serviceInstalled', "Installed {0}", constants.serviceName));
+					break;
+				case Events.DOWNLOAD_START:
+					this.outputChannel.appendLine(localize('downloadingService', "Downloading {0}", args[0]));
+					this.outputChannel.append(localize('downloadingServiceSize', "({0} KB)", Math.ceil(args[1] / 1024).toLocaleString(vscode.env.language)));
+					this.statusView.text = localize('downloadingServiceStatus', "Downloading {0}", constants.serviceName);
+					break;
+				case Events.DOWNLOAD_PROGRESS:
+					let newDots = Math.ceil(args[0] / 5);
+					if (newDots > dots) {
+						this.outputChannel.append('.'.repeat(newDots - dots));
+						dots = newDots;
+					}
+					break;
+				case Events.DOWNLOAD_END:
+					this.outputChannel.appendLine(localize('downloadingServiceComplete', "Done downloading {0}", constants.serviceName));
+					break;
+				case Events.ENTRY_EXTRACTED:
+					this.outputChannel.appendLine(localize('entryExtractedChannelMsg', "Extracted {0} ({1}/{2})", args[0], args[1], args[2]));
+					break;
+				case Events.LOG_EMITTED:
+					if (args[0] >= LogLevel.Warning) {
+						this.outputChannel.appendLine(args[1]);
+					}
+					break;
+				default:
+					break;
+			}
+		};
+	}
+}
+
+class CustomOutputChannel implements vscode.OutputChannel {
```

Comment:
Can you get rid of it? You already create an output channel for the extension so it seems better to just use that.